### ISSUE TITLE
Handshake failed: signature verification failed

### ISSUE THREAD

**Issue by andrewgbliss** [mscdex/ssh2#488]
I am using the latest of ssh2 and ssh2-streams: 0.5.2 && 0.1.11

I am not sure if this is a real issue with this module or the server I am trying to connect to.

Here is the debug output:

DEBUG: Local ident: 'SSH-2.0-ssh2js0.1.11'
DEBUG: Client: Trying myhost.com on port 22 ...
DEBUG: Client: Connected
DEBUG: Parser: IN_INIT
DEBUG: Parser: IN_GREETING
DEBUG: Parser: IN_HEADER
DEBUG: Remote ident: 'SSH-2.0-9.99 FlowSsh: Bitvise SSH Server (WinSSHD)'
DEBUG: Outgoing: Writing KEXINIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:508,padLen:14,remainLen:504
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
DEBUG: Comparing KEXINITs ...
DEBUG: (local) KEX algorithms: ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
DEBUG: (remote) KEX algorithms: ecdh-sha2-1.3.132.0.10,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group14-sha1,gss-group14-sha1-toWM5Slw5Ew8Mqkay+al2g==,ext-info-s
DEBUG: KEX algorithm: ecdh-sha2-nistp256
DEBUG: (local) Host key formats: ssh-rsa,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521
DEBUG: (remote) Host key formats: rsa-sha2-512,rsa-sha2-256,ssh-rsa
DEBUG: Host key format: ssh-rsa
DEBUG: (local) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Client->Server ciphers: aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr,3des-ctr
DEBUG: Client->Server Cipher: aes128-ctr
DEBUG: (local) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Server->Client ciphers: aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr,3des-ctr
DEBUG: Server->Client Cipher: aes128-ctr
DEBUG: (local) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha1
DEBUG: Client->Server HMAC algorithm: hmac-sha2-256
DEBUG: (local) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha1
DEBUG: Server->Client HMAC algorithm: hmac-sha2-256
DEBUG: (local) Client->Server compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Client->Server compression algorithms: zlib,none
DEBUG: Client->Server compression algorithm: none
DEBUG: (local) Server->Client compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Server->Client compression algorithms: zlib,none
DEBUG: Server->Client compression algorithm: none
DEBUG: Outgoing: Writing KEXECDH_INIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:900,padLen:15,remainLen:896
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXECDH_REPLY
DEBUG: Checking host key format
DEBUG: Checking signature format
DEBUG: Verifying host fingerprint
DEBUG: Host accepted by default (no verification)
DEBUG: Verifying signature
DEBUG: Signature verification failed
DEBUG: Outgoing: Writing DISCONNECT (KEY_EXCHANGE_FAILED)

It returns with the error: Handshake failed: signature verification failed

When I try to just use the command line sftp "user"@"host" the prompt waits about 10 - 15 seconds and shows this message:

"This system is for the use of authorized users only.
Individuals using this computer system without
authority, or in excess of their authority, are subject
to having all of their activities on this system
monitored and recorded by system personnel. In the
course of monitoring individuals improperly using this
system, or in the course of system maintenance, the
activities of authorized users may also be monitored.
Anyone using this system expressly consents to such
monitoring and is advised that if such monitoring
reveals possible evidence of criminal activity, system
personnel may provide the evidence of such monitoring
to law enforcement officials."

Then it prompts for the password. I am able to enter the password and get into the shell on their server just fine.

Do I need to implement some kind of waiting listener? 

I am have tried to add the property "tryKeyboard" and implement the "keyboard-interactive" event but it never hits the event at all, it errors immediately.

**Comment by mscdex** [comment]
What version of the Bitvise server are you using? I just tried with the current version on their site and I could not duplicate the problem after a few dozen attempts. I also verified that the handshake algorithms, ciphers, etc. were the same as what you've shown.

**Comment by andrewgbliss** [comment]
The BitVise version is 7.12. Here is my code:

const Client = require('ssh2').Client;

var conn = new Client();
conn.on('ready', () => {
  console.log('ready');
  conn.sftp((err, sftp) => {
    if (err) {
      console.log('ERROR', err);
      return err;
    } else {
      console.log('Got here');
      return sftp;
    }
  });
})
.on('error', err => {
  console.log('ERROR 2', err);
})
.connect({
  host: '****',
  port: 22,
  username: '****',
  password: '****',
  debug: console.log
});

And here is where the error is happening:

DEBUG: Signature verification failed
DEBUG: Outgoing: Writing DISCONNECT (KEY_EXCHANGE_FAILED)
ERROR 2 { Error: Handshake failed: signature verification failed
    at onKEXDH_REPLY (/app/node_modules/ssh2-streams/lib/ssh.js:2839:15)
    at SSH2Stream.<anonymous> (/app/node_modules/ssh2-streams/lib/ssh.js:212:45)
    at emitOne (events.js:96:13)
    at SSH2Stream.emit (events.js:188:7)
    at parse_KEXDH_REPLY (/app/node_modules/ssh2-streams/lib/ssh.js:4127:8)
    at parse_KEX (/app/node_modules/ssh2-streams/lib/ssh.js:4089:14)
    at parsePacket (/app/node_modules/ssh2-streams/lib/ssh.js:3931:12)
    at SSH2Stream._transform (/app/node_modules/ssh2-streams/lib/ssh.js:652:13)
    at SSH2Stream.Transform._read (_stream_transform.js:167:10)
    at SSH2Stream._read (/app/node_modules/ssh2-streams/lib/ssh.js:236:15) level: 'protocol' }

**Comment by andrewgbliss** [comment]
We have upgraded the BitVise server to the latest 7.15 and it is still giving the same error.

**Comment by mscdex** [comment]
Out of curiosity, what version of node are you using?

**Comment by andrewgbliss** [comment]
I am using v6.3.0 of node on my local mac. On my server I am using v6.3.1. Same results on both

**Comment by mscdex** [comment]
Would it be possible to connect to this particular server? I wouldn't need any credentials since it's a handshake error. If so, you can send the connection details to mscdex at mscdex.net.

**Comment by denisbider** [comment]
Hello,

this is denis from Bitvise. We have been contacted by a user - I think a different user - who is experiencing this issue. This user is not using ssh2js directly. Instead, ssh2js is being used by a high-profile website. This website is reporting the error described in this issue when their automated client connects to the user's SSH Server.

If you have access to a Windows computer or virtual machine - desktop or server, x86 or x64, all versions from Windows XP and 2003 until the most recent are supported - you can download Bitvise SSH Server free of charge here:

https://www.bitvise.com/ssh-server-download

Alternately, if you are not in a position to install it yourself, I have set up the following versions on an experiment server:

**Bitvise SSH Server version 7.15** - current most recent release
Host: experiment.bitvise.com
Port: 10715
RSA host key fingerprint: EDytg1YRiyMzbuReXIU5LFAiOsNqYM4xOin3OS8FGUI
ECDSA/nistp256 fingerprint: o++A9GlYCk+CLUtuCaWPWu/W6fp9buLlAitcGBHdYDM

**Bitvise SSH Server version 7.12** - as used by the user in this issue
Host: experiment.bitvise.com
Port: 10712
RSA host key fingerprint: /XnmtrmboAnt7JBR/5kCcXFE9+GEDfdI74kRVPs9P+k
ECDSA/nistp256 fingerprint: 286R8EeuHGpNkENONUxPhWsGYXT6kcaJBaq84HCZGUs

The errors reported by these users appear to be coming from here:

https://github.com/mscdex/ssh2-streams/blob/master/lib/ssh.js#L2903
var err = new Error('Handshake failed: signature verification failed');

I will subscribe to receive notifications about this issue. If you need to contact us for any other reason in the future, please feel free to use the Contact page on our website, which leads to a contact web form. You will *not* need to make a purchase, and at this time, there is no need to open any type of account, or a login.

**Comment by mscdex** [comment]
@denisbider I would have tried to install older versions to help narrow things down but I could not find links to older versions on the Bitvise website and the Amazon link used to host downloads does not allow directory listings.

The thing is, @andrewgbliss said he also tried 7.15 and ran into the same problem. With my 7.15 installation, I have no problems using the same node and `ssh2` module versions and key exchange algorithms.

**Comment by mscdex** [comment]
I am able to connect to the two Bitvise hosts mentioned by @denisbider without any problems.

@andrewgbliss can you try to connecting to both of those hosts and verify that they also connect for you (you will probably need debug output enabled to verify this as user authentication won't be successful)?

**Comment by andrewgbliss** [comment]
Yes I am able to connect to those servers.  It is fine when it verifies the signature.

**Comment by denisbider** [comment]
Hello Andrew,

thank you for confirming this. If I understand correctly, then:

- You can use your Node.js and ssh2-streams installation to connect to experiment.bitvise.com - both to the 7.12 and the 7.15 version?

- However: using the same Node.js and ssh2-streams installation, you encounter the reported error when you connect to another Bitvise SSH Server 7.12 installation?

Is this correct? Can you confirm that the same Node.js installation is involved?

Can you confirm if you have tested by negotiating both the RSA host key, and also the ECDSA host key? It is possible that the signature verification issue could be arising only with one algorithm, but not another.

If the above is correct, and you have tested both host key algorithms: then there is likely some important difference on the server, or some other aspect that is specific to the connection.

Do you have information about the Windows version on which Bitvise SSH Server runs in the case where you experience the signature verification issue?

Do you have administrative access to the involved SSH Server installation? In this case, are you able to generate a new host key for the SSH Server, using the same algorithm, and test using the re-generated host key again?

This would be to test the possibility that signature verification fails with some host keys, but not others, depending on some random property of the host key.

If the error still continues to happen with a re-generated host key, it would be necessary to repeat (re-generate again and test) another few times to make sure that the issue is not related to the specific host key that was generated.

**Comment by denisbider** [comment]
Another possible hypothesis is that the issue could be arising from a race condition, which might be triggered only if the latency between the server and client is low. For this purpose, can you check the ping time from your Node.js installation to experiment.bitvise.com; and compare the ping time from your Node.js installation to the SSH Server installation where signature verification fails?

**Comment by andrewgbliss** [comment]
The server is running Windows 2012 R2. Both of the Node.js scripts that I run fail instantly. I am not sure if there is any difference in ping time.

**Comment by denisbider** [comment]
I appreciate that information. The server experiment.bitvise.com runs Windows Server 2012 R2 as well, so there appears to be no difference there.

At this point, I can think of the following avenues to still investigate:

- If you have administrative access to the SSH Server installation, generate a new host key for the SSH Server, using the same algorithm, and test using the re-generated host key again. If it still fails, repeat the process multiple times with new keys, to ensure it appears to fail with all host keys generated.

- Measure latency (using "ping") from the computer running Node.js to the computer running the SSH Server installation where signature fails. Secondly, measure latency from the computer running Node.js to experiment.bitvise.com. It would be important to know both latencies.

We need to conclusively determine what difference there exists between experiment.bitvise.com and your SSH Server installation that correlates with the issue you're experiencing. At this point, we have ruled out OS version, since it appears it's the same on both installations.

**Comment by krazar** [comment]
Hello,

We have the same message while trying to connect to a Bitvise server.
client OS: ubuntu 16,  node: v6.9.1

This problem appears while upgrading ssh2 from 0.4.6 to last version.
We will stick to 0.4.6 fow now.

**Comment by mscdex** [comment]
@krazar What version of `ssh2-streams` did you have installed with `ssh2` v0.4.6? Unfortunately I encounter the same issue with `ssh2` v0.4.6 and `ssh2-streams` v0.0.23.

**Comment by mscdex** [comment]
@andrewgbliss Can you add an ECDSA key at least temporarily to this server?

**Comment by haomaggiewu** [comment]
How to output these debug message?

**Comment by mscdex** [comment]
@haomaggiewu `debug: console.log` in the object passed to `connect()`

**Comment by CodingPapi** [comment]
same issue when trying to connect to a Bitvise host on Windows 2008 R2 from fedora

**Comment by mscdex** [comment]
Closing this for now. If someone is able to either submit a PR for a fix *or* come up with instructions on how this issue can be reliably reproduced locally, please let me know. Until then there is nothing I can do.

**Comment by sanmai-NL** [comment]
What is the point of closing the issue?

**Comment by mscdex** [comment]
@sanmai-NL As I previously explained, there is nothing that can be done here until either of the two mentioned events occur (a PR for a fix and/or a way for me to reproduce this locally so I can fix it).

**Comment by sanmai-NL** [comment]
Should open issues always be immediately actionable?

**Comment by mscdex** [comment]
@sanmai-NL This issue has been open for some time and I previously spent a great deal of time trying to reproduce and debug the issue locally on my end. There hasn't been any real activity on this issue since November of last year. Given all of this, I don't see the point of leaving such an inactive issue open.

Again, I am more than willing to try and resolve whatever this issue is, but I cannot do so until I can fully debug the problem. This is very difficult to do with a closed-source product, especially with one that I cannot get to reproduce the issue locally.

**Comment by denisbider** [comment]
Closed source is not an issue. The SSH Server is freely available to test (including the public test server we provide).

The main obstacle is reproducing the issue. If the issue cannot be easily reproduced, I'm also not sure what to do.

**Comment by kryysler** [comment]
I seem to be having the same problem:

````
DEBUG: Local ident: 'SSH-2.0-ssh2js0.1.19'
DEBUG: Client: Trying IP on port PORT ...
DEBUG: Client: Connected
DEBUG: Parser: IN_INIT
DEBUG: Parser: IN_GREETING
DEBUG: Parser: IN_HEADER
DEBUG: Remote ident: 'SSH-2.0-9.99 FlowSsh: Bitvise SSH Server (WinSSHD)'
DEBUG: Outgoing: Writing KEXINIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:404,padLen:16,remainLen:400
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
DEBUG: Comparing KEXINITs ...
DEBUG: (local) KEX algorithms: ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
DEBUG: (remote) KEX algorithms: ecdh-sha2-1.3.132.0.10,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1,gss-group14-sha1-toWM5Slw5Ew8Mqkay+al2g==
DEBUG: KEX algorithm: ecdh-sha2-nistp256
DEBUG: (local) Host key formats: ssh-rsa,ssh-dss
DEBUG: (remote) Host key formats: ssh-dss
DEBUG: Host key format: ssh-dss
DEBUG: (local) Client->Server ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Client->Server ciphers: aes256-ctr,3des-ctr,aes256-cbc,3des-cbc
DEBUG: Client->Server Cipher: aes256-ctr
DEBUG: (local) Server->Client ciphers: aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com
DEBUG: (remote) Server->Client ciphers: aes256-ctr,3des-ctr,aes256-cbc,3des-cbc
DEBUG: Server->Client Cipher: aes256-ctr
DEBUG: (local) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Client->Server HMAC algorithms: hmac-sha2-256,hmac-sha1
DEBUG: Client->Server HMAC algorithm: hmac-sha2-256
DEBUG: (local) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG: (remote) Server->Client HMAC algorithms: hmac-sha2-256,hmac-sha1
DEBUG: Server->Client HMAC algorithm: hmac-sha2-256
DEBUG: (local) Client->Server compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Client->Server compression algorithms: zlib,none
DEBUG: Client->Server compression algorithm: none
DEBUG: (local) Server->Client compression algorithms: none,zlib@openssh.com,zlib
DEBUG: (remote) Server->Client compression algorithms: zlib,none
DEBUG: Server->Client compression algorithm: none
DEBUG: Outgoing: Writing KEXECDH_INIT
DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
DEBUG: Parser: IN_PACKET
DEBUG: Parser: pktLen:580,padLen:13,remainLen:576
DEBUG: Parser: IN_PACKETDATA
DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXECDH_REPLY
DEBUG: Checking host key format
DEBUG: Checking signature format
DEBUG: Verifying host fingerprint
DEBUG: Host accepted by default (no verification)
DEBUG: Verifying signature
DEBUG: Signature verification failed
DEBUG: Outgoing: Writing DISCONNECT (KEY_EXCHANGE_FAILED)
````

Environment:
````
NodeJS@6.10
ssh2@0.5.5
ssh2-streams@0.1.19
````

I wasn't able to connect to the Bitvise 7.15 test server posted above in this thread, it just timed out. probably not available anymore.

I'm not in control of the SSH server I'm trying to connect to, so don't have any more information, but I can ask the provider if there are some specifics that would be important to resolve the issue.

Any help would be appreciated.

**Comment by denisbider** [comment]
@kryysler, on Bitvise's behalf, thank you for the update.

I have restored access to the 7.12 and 7.15 test installations described above. In addition, the following additional installations are now also available:

Bitvise SSH Server version 7.28
Host: experiment.bitvise.com
Port: 10728
RSA host key fingerprint: uOkdMKvJTY45Orop3R5GcW4p9LeOns87bPt6t92yjmI
ECDSA/nistp256 fingerprint: bamClRZS5l9v1YGnuUz7ienTuoTelhq6kfVG/mM4UX0

Bitvise SSH Server version 7.32 - current most recent release
Host: experiment.bitvise.com
Port: 10732
RSA host key fingerprint: RhAYcAMnpNP+jmQ88NZ713fxJpImQhKuNAgtWVT5fdc
ECDSA/nistp256 fingerprint: 1C1JdwjDTsdZy6A4u7N47TDJ9nQzZoQyYVl3jDnwUJ8

If you need me to add any additional host keys (e.g. DSA), please let me know.

**Comment by aart-janboor** [comment]
@denisbider was this ever solved? I now also ran into the this issue. Funny thing is we have two Bitvise SFTP servers (same version, most current one) and on 1 server we don't run into this issue while in another we do. I traced it back to var verified = verifier.verify.... in the ssh2 lib code and against both servers it seems to be initiated with similar params (different keys of course). Even after regeneration of the keys on the server the issue persists. My reasoning is as follows;

1. It's unlikely its something specific to NodeJS as we have an application working against 1 Bitvise SFTP server just fine whilst another gives this error.
2. It's also unlikely to be a very apparent issue in bitvise as e.g. we can connect with Filezilla to both SFTP servers just fine. 
3. So it seems it must have something to do with the handshake between ssh2 lib and the specific instance of the Bitvise server where something is going wrong. My knowledge of the subject is insufficient to troubleshoot this any deeper.

**Comment by mscdex** [comment]
@aart-janboor IMHO the problem in the past has not been finding a server exhibiting the behavior (IIRC andrewgbliss provided me with connection details early on), but getting (sufficient) debug output from the Bitvise server to see what exactly is going on, especially when it comes to calculating the shared secret and such. Without that kind of information it's pretty much impossible to find the root cause.

**Comment by aart-janboor** [comment]
@mscdex OK I understand , thanks for the feedback. Is there any way we could use Filezilla's source code to troubleshoot the issue - as the Filezilla client has no issue in connecting to the troubled server - or is there any other way I could be of help?

**Comment by mofux** [comment]
I've never used Bitvise, but maybe it would be possible to swap the configuration and keys of both servers and see if the problem moves with the config. If it does we can narrow it down on the config, otherwise it would point to some problem with the environment the Bitvise server runs in...

**Comment by denisbider** [comment]
@aart-janboor - if you can reproduce the issue with a Bitvise SSH Server installation under your control, the SSH Server can be configured to perform extremely detailed logging that will capture everything sent and received, as well as internal events in the server.

To turn on Debug logging:

- First, stop the SSH Server. Try to ensure that no connections, except the one you are interested in, will occur while Debug logging is enabled.

- Open Advanced SSH Server Settings.

- Under Logging, set "Textual log file logging level" to "Custom events".

- Click on "Textual log file events" to open an event selection dialog. Make sure that "Log Errors", "Log Warnings", "Log Info", and "Log Trace" are all CHECKED.

- Make sure that "Log Debug" is UNCHECKED - we don't want ALL the possible Debug log messages.

- Sort the event list by Type until Debug is on top. Enable just this one event: "D_FLOW_SSH_MESSAGE_SENT".

- Restart the SSH server. This will create a new log file.

- Reproduce ONE connection that results in the issue.

- Stop the SSH Server so that it will close the log file.

- Set your textual log file settings back to normal (Errors, Warnings, Info) before restarting the SSH Server.

Log files are recorded by default in the Logs subdirectory of the SSH Server installation directory. After following the above steps, the debug log file will be larger than usual, and will contain detailed activity of the SSH session.

Since the signature verification is failing on the client - not the server - ultimately the entire exchange will likely need to be investigated by the client developers. However, I can help analyze the log file and extract information that would help the client developers.

If you want to send me the Debug log file for analysis, the best way is through our case management system:

https://fogbugz.bitvise.com/?pg=pgPublicEdit

**Comment by denisbider** [comment]
FYI - I have heard of a similar issue (a signature verification failure) happening with a commercial implementation - Attachmate Reflection (now Micro Focus). In that case, the information I have is that they identified it as a bug in their software, and fixed it in a subsequent version. I don't have details of what exactly they fixed.

**Comment by denisbider** [comment]
@mscdex - if the SSH Server is configured for debug logging as described above, it should contain way enough information including what's needed for secret calculations. If you get a debug log file for an affected connection, and would like me to help extract the needed info, let me know.

**Comment by mscdex** [comment]
@denisbider That's good to know and hear. Thanks.

**Comment by aart-janboor** [comment]
@denisbider @mscdex thanks for the feedback both - I'll see what I can do to help further narrow down this issue based on your input. I just installed a second bitvise instance on the troubled server and with vanilla config it doesn't exhibit the issue. That is promising because it implies I might be able to narrow it down to the activation of a specific config parameter/feature on the server.

**Comment by aart-janboor** [comment]
![server version](https://user-images.githubusercontent.com/11685254/30364108-1dbb377e-9863-11e7-8610-81249829bf1f.png)

@denisbider @mscdex @mofux  With help of your pointers I've been able to resolve the issue for our server :-) It turned out enabling the option "Omit server version" was the culprit. We had enabled that option for security hardening but disabling it makes the signature verification failed issue go away.

For now we can live with this workaround. At the same time I hope that this information helps in finding the ultimate root cause of the issue and solving it either on the Bitvise side or on the SSH side permanently.

**Comment by denisbider** [comment]
@aart-janboor Wow, very interesting.

@mscdex - what it does is, it just changes the way the SSH version string is constructed. There are no other changes to the flow of the SSH session. Example of version string without this setting:

    53 53 48 2d 32 2e 30 2d  37 2e 33 34 20 46 6c 6f   SSH-2.0-7.34 Flo
    77 53 73 68 3a 20 42 69  74 76 69 73 65 20 53 53   wSsh: Bitvise SS
    48 20 53 65 72 76 65 72  20 28 57 69 6e 53 53 48   H Server (WinSSH
    44 29 20 37 2e 33 34 0d  0a                        D) 7.34..

Example of version string with this setting:

    53 53 48 2d 32 2e 30 2d  39 2e 39 39 20 46 6c 6f   SSH-2.0-9.99 Flo
    77 53 73 68 3a 20 42 69  74 76 69 73 65 20 53 53   wSsh: Bitvise SS
    48 20 53 65 72 76 65 72  20 28 57 69 6e 53 53 48   H Server (WinSSH
    44 29 20 0d 0a                                     D) ..

Notice the single space at the end.

You have a bug in that you are trimming that space, and this corrupts the way your exchange hash is calculated.

We've seen this bug in other software previously, so I'll see what I can do to have that space trimmed on our end.

**Comment by denisbider** [comment]
@mscdex - Actually... I'm not sure it's a good idea to change this on our end.

The version string is valid as-is, with or without the space at the end. But if I remove that space in new versions, it will make it easy to distinguish that new version from previous ones, which defeats the purpose of the setting.

**Comment by aart-janboor** [comment]
@denisbider I would actually propose you improve Bitvise. Frankly speaking from security hardening perspective the feature could be further improved by making sure it does not disclose the fact it's a Bitvise server at all and the version string is either empty or can be customized. I think the additional value of this enhancement offsets the limited information it provides to a potential attacker - as it would still not be able to derive the exact version of the SSH server. You would only be able to know if the version precedes the version where this enhanced security feature was introduced but not the exact version number. 

As a bonus you prevent this specific issue from surfacing again - as you've now already seen two client tools that exhibit this behaviour and trimming inputs like these is not super uncommon you'll likely face it again with other client libraries in the future if you don't change it.

**Comment by mofux** [comment]
Now that we know what it is, this issue was already reported on the ssh2streams dependency: https://github.com/mscdex/ssh2-streams/issues/72 and proposed to be fixed via the pending PR https://github.com/mscdex/ssh2-streams/pull/73

**Comment by denisbider** [comment]
@aart-janboor - unfortunately that reasoning is unsound. There is no discernible security benefit from hiding the product. Q550 in our SSH Server Usage FAQ explains why:

https://www.bitvise.com/ssh-server-usage-faq#Q550

An attacker can use nmap to narrow down your OS to a specific Windows version range, based on mere behavior of the TCP stack. No need for the SSH Server to send a version string at all. Then all the attacker needs to do is look at the structure of the KEXINIT packet, and easily see what kind of server it is dealing with, again with no version string at all.

**Comment by aart-janboor** [comment]
@denisbider thanks for providing that insight. Interestingly enough you mention in your FAQ " KEXINIT packets are sent in plaintext and have specific patterns which are sufficient not just to identify the make of the server, but also a particular version subset." which seems to suggest the feature itself it 'useless' in the sense that even the version number itself can be derived from the KEXINIT packet. Based on that and your OpenSSH remark perhaps worth considering to just completely get rid of the feature in Bitvise which would also solve the issue and also reduce false sense of security.

**Comment by denisbider** [comment]
@aart-janboor - there is some security in the feature because minor versions may fix issues, but not involve changes in KEXINIT or in fact at any point during key exchange before authentication. So with "Omit server version" enabled, an attacker can determine this is an SSH Server version 7.2x, but not which "x".

So there is some value in it, and that's why I'm considering to postpone removing the trailing whitespace until the next significant feature version. If we remove the trailing whitespace in 7.35, then it will be possible to distinguish it from 7.34, which would otherwise not be the case. But if we do this in a significant feature version, that's likely to contain other changes that are fingerprintable anyway.

**Comment by mscdex** [comment]
@mofux That PR is only for the ident being sent *from* `ssh2-streams`. A separate change will be needed for the *incoming* ident from the remote party.

**Comment by mscdex** [comment]
This should be fixed in mscdex/ssh2-streams@09c5884eccd7. Please try `npm install`'ing the master branch of `ssh2-streams` and let me know if that does fix it for you @aart-janboor.

@andrewgbliss @CodingPapi @kryysler If you any/all of you could try this as well and see if that solves the issue for you as well that would be good to know.
